test/MTLSpec.hs

```diff
@@ -0,0 +1,143 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module MTLSpec where
+
+import           Polysemy
+import           Polysemy.Reader
+import           Polysemy.Writer
+import           Polysemy.State
+import           Polysemy.Error
+import           Polysemy.MTL
+
+import qualified Data.Text                     as T
+import           Test.Hspec
+
+import           Control.Monad                  ( replicateM_
+                                                , when
+                                                )
+import qualified Control.Monad.Reader.Class    as S
+import qualified Control.Monad.Writer.Class    as S
+import qualified Control.Monad.State.Class     as S
+import qualified Control.Monad.Error.Class     as S
+
+
+{-
+We could re-write these to use polysemy directly.  Imagine, though
+that these come from external libraries so you can't so easily
+re-write them.
+-}
+getEnvLength :: S.MonadReader T.Text m => m Int
+getEnvLength = S.ask >>= return . T.length
+
+replicateTell :: S.MonadWriter [Int] m => Int -> Int -> m ()
+replicateTell n m = replicateM_ n $ S.tell [m]
+
+retrieveAndUpdateN :: S.MonadState Int m => Int -> m Int
+retrieveAndUpdateN n = do
+  m <- S.get
+  S.put n
+  return m
+
+-- this one is exceptionally boring
+throwOnZero :: S.MonadError T.Text m => Int -> m Int
+throwOnZero n = do
+  when (n == 0) $ S.throwError "Zero!"
+  return n
+
+someOfAll
+  :: (S.MonadReader T.Text m, S.MonadWriter [Int] m, S.MonadState Int m)
+  => m T.Text
+someOfAll = do
+  n <- S.get
+  S.tell [n]
+  S.ask
+
+spec :: Spec
+spec = do
+  describe "MTL" $ do
+
+    it
+        "should return 10, the sum of lengths of the strings provided to run (\"Text1\") and then to local (\"Text2\")"
+      $ do
+          flip shouldBe 10 $ run . runReader "Text1" $ do
+            a <- absorbReader getEnvLength
+            b <- local (const "Text2") $ absorbReader getEnvLength
+            return (a + b)
+
+    it
+        "should return the sum, after censoring, of all things told.  In this case, 16, the sum of \"init [1,5,5,5,5]\""
+      $ do
+          flip shouldBe 16 $ sum . fst . run . runWriter $ do
+            tell [1]
+            absorbWriter $ replicateTell 2 5
+            censor init $ absorbWriter $ replicateTell 2 5
+
+    it "Should return 0 (since 10 - (20 `div` 2) = 0)"
+      $ flip shouldBe 0
+      $ fst
```

Comment:
Here and elsewhere, I try to make the first `n - 1` function applications be `.`, and then the last one a `$`. Sometimes rank-n-edness doesn't allow this to happen, but prefer it when it does.

Reply:
I did what I could here.  There's something going on that makes it not work always so I had to put some ```$``` back in.